doc: add draft publish guideline (#178)
nodejs/package-maintenance: docs/drafts/PUBLISH-GUIDELINES.md

| publish config                 | [`publishConfig`]           | Helpful if you need to publish to a different registry    |
| publish package                | [`npm publish`]             | Actually publish the package                              |
| search/view published packages | [public registry]           | [for example](https://www.npmjs.com/package/npm)          |
| filter publish files           | [`.npmignore`] or [`files`] | Use one base on your preference                           |

## `.npmignore` or `files`

[npm] has default ignore rules and loads from `.gitignore` but it also offer two methods for you to keep your published packages tidy.

- blacklisting - specifying ignore files in `.npmignore`
- whitelisting - specifying files you only want to publish with `files` in your `package.json`.

Which one to go with is solely up to you and your preference, but beware of the following gotchas and tips:

On the `.gititnore` file:

- [npm] automatically uses `.gitignore` if neither one exist

On the `.npmignore` file:

- using `.npmignore`, you could potentially publish any unintended files left in the directory.

> - [npm] will stop using `.gitnore` if this exist, so beware because you may forget to add some rules to `.npmignore` after you add them to `.gitignore`

Reviewer: - - [npm] will stop using `.gitnore` if this exist, so beware because you may forget to add some rules to `.npmignore` after you add them to `.gitignore`
+ - [npm] will stop using `.gitnore` if this exists, so beware because you may forget to add some rules to `.npmignore` after you add them to `.gitignore`

> - [npm] will stop using `.gitnore` if this exist, so beware because you may forget to add some rules to `.npmignore` after you add them to `.gitignore`

Reviewer: ```suggestion
- [npm] will stop using `.gitnore` if this exists, so beware because you may forget to add some rules to `.npmignore` after you add them to `.gitignore`
```

> - [npm] will stop using `.gitnore` if this exist, so beware because you may forget to add some rules to `.npmignore` after you add them to `.gitignore`

Reviewer 2: ```suggestion
- [npm] will stop using `.gitignore` if this exists, so beware because you may forget to add some rules to `.npmignore` after you add them to `.gitignore`
```

On the `files` field in `package.json`:

- the opposite of `.npmignore`, if you forget to add some new files to this, you could publish a package that's broken because of missing files.

> - there's an open bug that [`files` affects npm's default rules](https://npm.community/t/ds-store-files-show-up-after-npm-publish/831/4)

Reviewer: It might be better to expand on what that means to the reader versus them having to go an read the reference.

## Verification

> In the newer [npm] version 6, [`npm publish`] now shows the list of files published, but it doesn't offer you a chance to review them first.

Reviewer 2: Add `--dry-run` entry after this line
```suggestion
In the newer [npm] version 6, [`npm publish`] now shows the list of files published.
```

> If you want to review the files, you can use the [`npm pack`] command to generate a tarball first and look inside it.

Reviewer 2: ```suggestion
If you want to review the files, you can use the [`npm pack`] command to generate a tarball first and look inside it or run `npm publish --dry-run` to get the file list.
```

You can also use the [`npm pack`] command to download a tarball from a published version.

For example: `npm pack npm@6.5.0`.

## Sources, Tests, and Docs

> A package not written in idiomatic JavaScript should be published with transpiled JavaScript as the primary executable.

Reviewer: It might be worth expanding on "not written in idiomatic JavaScript"

| bump version                   | [`npm version`]             | This bumps version and creates git commit and tag for you |
| preview/view tarball           | [`npm pack`]                | Allows you to see files before or after publishing        |
| publish config                 | [`publishConfig`]           | Helpful if you need to publish to a different registry    |
| publish package                | [`npm publish`]             | Actually publish the package                              |
| search/view published packages | [public registry]           | [for example](https://www.npmjs.com/package/npm)          |
| filter publish files           | [`.npmignore`] or [`files`] | Use one base on your preference                           |

## `.npmignore` or `files`

[npm] has default ignore rules and loads from `.gitignore` but it also offer two methods for you to keep your published packages tidy.

- blacklisting - specifying ignore files in `.npmignore`
- whitelisting - specifying files you only want to publish with `files` in your `package.json`.

Which one to go with is solely up to you and your preference, but beware of the following gotchas and tips:

> On the `.gititnore` file:

Reviewer 2: ```suggestion
On the `.gitignore` file:
```
